added traversal algorithm to nx_parallel

RohitP2005 opened the pull request:
I added traversal algorithm with dfs and bfs which is ```_dispatchable``` . Kindly suggest further corrections

RohitP2005 commented:
@stefanv can u help with what to be done next

dschult reviewed (commented):
This is a lot of new functions!  
But I can't tell if it is doing anything that networkx is not already doing.
For example the `breadth_first_search.py` functions look copied straight fro networkx. That code is not parallel. I think the idea is to split up the "embarrasing parallelizable" loop and do the iterations of that loop on different processors. 

Maybe just focus on one function and get that working... Then you can do the others.
To get an idea of how this works look at the other functions already in nx-parallel. For example, `betweenness_centrality`.

RohitP2005 commented:
> This is a lot of new functions! But I can't tell if it is doing anything that networkx is not already doing. For example the `breadth_first_search.py` functions look copied straight fro networkx. That code is not parallel. I think the idea is to split up the "embarrasing parallelizable" loop and do the iterations of that loop on different processors.
> 
> Maybe just focus on one function and get that working... Then you can do the others. To get an idea of how this works look at the other functions already in nx-parallel. For example, `betweenness_centrality`.

Sorry about that , i will look into it asap , thanks for the help

RohitP2005 commented:
@dschult
Thanks so much for your detailed feedback! You're absolutely right—it’s better to focus on implementing and testing one function thoroughly before moving on to others.

I’ve removed the DFS algorithm for now and added a BFS algorithm that is parallelized for improved performance.

Here’s an example of how to use the new parallel_bfs function:
```
# Create a sample graph
G = nx.erdos_renyi_graph(100, 0.05)

# Perform parallel BFS
bfs_result = parallel_bfs(G, source=0)
print("BFS Traversal Order:", bfs_result)

```

Schefflera-Arboricola reviewed (commented):
Thanks @RohitP2005 for the PR! Could you please follow this checklist for adding a new algorithm to the project - https://github.com/networkx/nx-parallel/blob/main/CONTRIBUTING.md#general-guidelines-on-adding-a-new-algorithm and you can also look at the existing algorithms in the repository to get a better idea! If any of these points in the checklist are not clear then please let us know. Thanks!

RohitP2005 commented:
@Schefflera-Arboricola Yeah sure , i will look into it.
I will make sure the new algorithm added meets your expectations

Schefflera-Arboricola commented on nx_parallel/algorithms/traversal/breadth_first_search.py:
```diff
@@ -0,0 +1,96 @@
+
+@nxp._configure_if_nx_active()
+@py_random_state(3)
+def parallel_bfs(G, source=None, get_chunks="chunks", n_jobs=None):
```
n_jobs is included in the configurations. so no need to include it as a parameter here. (see - https://github.com/networkx/nx-parallel/blob/main/Config.md)

Schefflera-Arboricola replied:
the names of the functions in the main networkx project and in this nx-parallel project are usually kept same. (see - https://github.com/networkx/nx-parallel/blob/main/CONTRIBUTING.md#general-guidelines-on-adding-a-new-algorithm)

Schefflera-Arboricola commented on nx_parallel/algorithms/traversal/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from .depth_first_search import *
```
```suggestion
```

Schefflera-Arboricola commented on nx_parallel/algorithms/traversal/breadth_first_search.py:
```diff
@@ -0,0 +1,96 @@
+
+    Parameters
+    ----------
+    G : graph
+        A NetworkX graph.
+    source : node, optional
+        Starting node for the BFS traversal. If None, BFS is performed for all nodes.
```
you don't need to document the parameters that are already documented in the main networkx repo :)

Schefflera-Arboricola reviewed (commented):
Thanks for all the work @RohitP2005 !

I haven't looked at the core implementation in this PR too closely.. but since BFS is not an embarrassingly parallel algorithm, I don't think its parallel implementation could look this simple 😅. If you are new to parallel graph algorithms then I'd suggest you consider adding an embarrassingly parallel algorithm instead (see - https://github.com/networkx/nx-parallel/issues/82). BFS for starters might be not be great.

Thanks, and sorry for the delayed reviews :)
